### PR TITLE
Added vaccine design script and unit test

### PR DESCRIPTION
**Pull request by jneich**
This script takes a protein fasta as an input and returns the optimal ordering of peptides to minimize creation of problematic neoepitopes, including spacers if necessary. 

Important dependencies, which are not already included in pvacseq are:
- pandas
- biopython
- networkx
- simanneal 

Biopython is used to parse the input fasta. 

IEDB output is stored in a pandas dataframe for filtering.

Networkx is used to construct a graph with peptide nodes. Edge attributes correspond to IC50 scores between these nodes and hold spacer information if necessary.

The 'longest path' (best ordering based on junction IC50 scores) is found using an overloaded function from the simanneal package (https://github.com/perrygeo/simanneal). 

This pull request also includes a unittest, which runs 10 iterations to determine if test files match expected output. This is because simulated annealing relies partly on chance to determine if a particular edge should be included in a path, so peptide ordering of sequential runs may not match identically.

### CODE REVIEW

**Comment by chrisamiller** [comment]
Right off the bat, you should figure out why the automated tests fail on your version.

File: pvacseq/lib/vaccine_design.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+             "Multiple epitope lengths can be specified " +
+             "using a comma-separated list. " +
+             "Typical epitope lengths vary between 8-11. " +
+             "Required for Class I prediction algorithms",
+    )
+    parser.add_argument("-c", "--cutoff", type=int,
+                        default=500,
+                        help="Optional ic50 cutoff value." +
+                             " Neoepitopes with IC50 values below this " +
+                             " value will be labeled as problematic. Default: 500")
```
Comment (chrisamiller) [inline pvacseq/lib/vaccine_design.py]: suggested edit: Junctional neoepitopes with scores below this ic50 value will be excluded.

File: pvacseq/lib/vaccine_design.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    parser.add_argument(
+        "-r", "--iedb-retries", type=int,
+        default=5,
+        help="Number of retries when making requests to the IEDB RESTful web interface. " +
+             " Must be less than or equal to 100." +
+             "Default: 5"
+    )
```
Comment (chrisamiller) [inline pvacseq/lib/vaccine_design.py]: Does this also inherit the argument that gives the option to run IEDB from a local install? If not, it probably needs to be added.

File: pvacseq/lib/vaccine_design.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    for ID_1 in Paths:
+        try:
+            distance_matrix[ID_1]
+        except KeyError:
+            distance_matrix[ID_1] = {}
+        for ID_2 in Paths[ID_1]:
+            distance_matrix[ID_1][ID_2] = Paths[ID_1][ID_2]['weight']
+
+    init_state = seq_keys
+    #random.shuffle(init_state)      # This line helps, but makes unit testing very difficult
```
Comment (chrisamiller) [inline pvacseq/lib/vaccine_design.py]: The usual way to get around this is to have the test provide a seed to the RNG so that the output is identical for test runs

File: pvacseq/pvacseq.py [hunk 4]
```diff
@@ -71,6 +71,11 @@ def main():
     config_files_parser.set_defaults(func=lib.config_files)
 
     parser.add_argument(
+        "vaccine_design",
+        help="Generate optimal vaccine configuration from protein fasta.",
```
Comment (chrisamiller) [inline pvacseq/pvacseq.py]: "optimal" is a loaded word. Maybe change to something doesn't necessarily imply that you're guaranteed to find the absolute best solution (because I don't believe that this algorithm does that).  Perhaps this?: "Generate an optimized ordering for peptides in a DNA vaccine"

File: tests/test_vaccine_design.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        self.assertTrue(compiled_script_path)
+
+    def test_vaccine_design_runs_and_produces_expected_output(self):
+        output_dir = tempfile.mkdtemp()
+        tempfile.tempdir = output_dir
+
+        if self.python is None:
+            raise Exception("Python could not be found!")
+
+        for i in range(0, 10):   # try this 10 times to make sure we pass once
```
Comment (chrisamiller) [inline tests/test_vaccine_design.py]: Why is this necessary?

**Comment by agraubert** [comment]
It looks like your tests failed from not including the new dependencies for the unit tests.  It's something that's easy to miss before uploading, so no worries.  You'll want to add the dependencies you listed above to [this list](https://github.com/griffithlab/pVAC-Seq/blob/master/setup.py#L44) which should fix the error that failed your tests yesterday

**Comment by agraubert** [comment]
Also, I'm going to change the base of the PR to our `develop` branch, which is where we put all our work before we're ready to release it.

**Comment by agraubert** [comment]
I've tested a rebase on a local branch.  If you run `git rebase --onto develop master vaccine_design`
then you only need to fix one conflict

File: pvacseq/lib/vaccine_design.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    from .. import lib
+except ValueError:
+    import lib
+
+import pandas
+import networkx as nx
+import itertools
+from Bio import SeqIO
+import lib
+from parse_output import *
```
Comment (agraubert) [inline pvacseq/lib/vaccine_design.py]: I don't think you ever use any of the functions from this file.  This is also causing issues since `parse_output.py` has been removed on the `develop` branch

File: pvacseq/pvacseq.py [hunk 7]
```diff
@@ -71,6 +71,11 @@ def main():
     config_files_parser.set_defaults(func=lib.config_files)
 
     parser.add_argument(
```
Comment (agraubert) [inline pvacseq/pvacseq.py]: This should be a subparser:
```python
vaccine_design_parser = subparsers.add_parser(
    "vaccine_design",
    help="Generate optimal vaccine configuration from protein fasta.",
    add_help=False
)
vaccine_design_parser.set_defaults(func=lib.vaccine_design)
```

File: tests/test_vaccine_design.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import subprocess
+from filecmp import cmp
+import os
+import sys
+
+#python -m unittest tests/test_vaccine_design.py
+class TestVaccineDesign(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        base_dir = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..'))
```
Comment (agraubert) [inline tests/test_vaccine_design.py]: `os.path.dirname(path)` is preferable to `os.path.join(path, '..')` mostly for consistency's sake

File: tests/test_vaccine_design.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        cls.python = sys.executable
+        cls.executable_dir = os.path.join(base_dir, 'lib')
```
Comment (agraubert) [inline tests/test_vaccine_design.py]: This should be `os.path.join(base_dir, 'pvacseq', 'lib')`

File: pvacseq/lib/vaccine_design.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
+#  Input would be a protein fasta
+# ie:
+# python vaccine_design.py test peptides.fa -o output.csv ann  H-2-Kb -e 8
+import shutil
+import sys
+from pathlib import Path
+root = str(Path(__file__).resolve().parents[1])
+sys.path.append(root)
+
+try:
```
Comment (agraubert) [inline pvacseq/lib/vaccine_design.py]: You can just directly `import lib`.  There's no need for the relative import

File: pvacseq/lib/vaccine_design.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+import networkx as nx
+import itertools
+from Bio import SeqIO
+import lib
+from parse_output import *
+
+import time
+from simanneal import Annealer
+import random
+import math
```
Comment (agraubert) [inline pvacseq/lib/vaccine_design.py]: You're missing a couple other imports like `argparse` and `os`

File: pvacseq/lib/vaccine_design.py [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+    parser = argparse.ArgumentParser('pvacseq vaccine_design')
+    parser.add_argument(
+        "run_name",
+        help="The name of the run being processed." +
+             " This will be used as a prefix for output files."
+    )
+    parser.add_argument('input_file', type=argparse.FileType('r'),
+                        help="Path to input FASTA file")
+    parser.add_argument('method',
+                        choices=PredictionClass.iedb_prediction_methods(),
```
Comment (agraubert) [inline pvacseq/lib/vaccine_design.py]: Careful when directly copy-pasting.  `PredictionClass` is not imported from anywhere in your code

**Comment by agraubert** [comment]
The rebase should have cleaned up this diff.  Why did dd28c4e re-introduce all these commits from master?

**Comment by jneich** [comment]
All the conflicts were outdated test data sets. There was only one conflict with code.

**Comment by agraubert** [comment]
Right, but I'm trying to avoid mixing commits from `master` and `develop`.  Our setup always involves keeping `develop` ahead of `master`, but this pr is now mixing in commits from `master` back into `develop`.  Most of the changes are benign, but one big issue is that this is re-introducing (parse_output.py)[https://github.com/jneich/pVAC-Seq/blob/vaccine_design/pvacseq/lib/parse_output.py] which has been deleted in the `develop` branch.  f0a4edf59506be99215a51f0917dd5a9585e0993 appears to be in the right state following a rebase, but then dd28c4e added the commits back.

**Comment by agraubert** [comment]
Oh, yeah, that's my bad.  The new dependencies will need to be added to `.travis.yml` but it looks like you've figured that out

**Comment by jneich** [comment]
Ok, I'll step back to `f0a4edf` then add the yml dependencies to that commit. Thanks!

**Comment by agraubert** [comment]
I'm still seeing `parse_output.py`.  Would you mind pushing a commit that directly `git rm`'s it?

File: setup.py [hunk 13]
```diff
@@ -41,7 +41,10 @@
         'PyVCF',
         'requests',
         'PyYAML',
-        'connexion'
+        'connexion',
+        'pandas',
+        'networkx',
```
Comment (agraubert) [inline setup.py]: Noticed this in my testing.  We need `simanneal` on this list too

Comment (jneich) [inline setup.py]: ah, of course

File: pvacseq/lib/vaccine_design.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+                             " Junctional neoepitopes with IC50 values below this " +
+                             " value will be excluded. Default: 500")
+    parser.add_argument(
+        "-r", "--iedb-retries", type=int,
+        default=5,
+        help="Number of retries when making requests to the IEDB RESTful web interface. " +
+             " Must be less than or equal to 100." +
+             "Default: 5"
+    )
+     parser.add_argument(
```
Comment (agraubert) [inline pvacseq/lib/vaccine_design.py]: I think your alignment is off by a single space here (just kinda eyeballing it from GH).  Travis is failing on this line due to an unexpected indentation.

**Comment by jneich** [comment]
The code has been updated and is ready to be integrated unless there are any final comments!

**Comment by susannasiebert** [comment]
I think it looks good. I would like to update the develop branch to pull in the latest changes from master. This should make it so that you don't have a bunch of my commits as part of this PR. So please hold off on merging this for now.

**Comment by susannasiebert** [comment]
@jneich Do you have permissions to merge this PR? Otherwise I can merge it for you.

**Comment by jneich** [comment]
@susannasiebert I do not have the permissions, so it would be great if you could take care of it. Thanks!
